### PR TITLE
Typescript naming rules

### PR DESCRIPTION
**Pull request by gmcgibbon**
## Description

Adds two naming rules:
- Ensure all generic type names start with a `T` prefix
- Ensure all interfaces _don't_ start with an `I` prefix

I think we could do with making a decision on these two naming conventions. These make sense to me for the following reasons:

#### The `T` prefix on generics:
- TypeScript is a Microsoft language, and `T` names and prefixes are in the [official docs](https://www.typescriptlang.org/docs/handbook/generics.html)
- Other MS languages like [C#](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-5.0) use this standard (my background when it comes to typed languages)
- Makes it easier to identify when I type is a generic, and not a concrete type filling a generic's place
- It allows people coming from other MS languages to onboard quicker to TypeScript
- I can’t find an example in the handbook where it doesn't just use T or some other letter which is slightly disheartening, but using the prefix is a visual compromise between clear generic type indication and human readable naming

#### The lack of `I` prefix on interfaces:
- Again, the TypeScript handbook does not do this, and we currently don't either (from what I've seen)
- This seems like a very _enterprise_ rule that might be a little antiquated
- This probably doesn't need as much selling, but having a linting rule that explicitly complains when others do this will help make code consistent

## Type of change

- [x] `@shopify/esling-plugin` Minor: New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have added a changelog entry, prefixed by the type of change noted above (Documentation fix and Test update does not need a changelog as we do not publish a new version for these changes)

### CODE REVIEW

**Comment by 9at8** [comment]
Makes sense to me

**Review by atesgoral — APPROVED** [review]

**Comment by kaelig** [comment]
Just to make sure I understand correctly the regexp, this still allows an interface to be called `Image` (starts with an `I`, followed by a lowercase character)?

File: packages/eslint-plugin/lib/config/typescript.js [hunk 1]
```diff
@@ -69,6 +69,23 @@ module.exports = {
         '@shopify/typescript/prefer-singular-enums': 'error',
         // Prefer buildClientSchema for schema building.
         '@shopify/typescript/prefer-build-client-schema': 'error',
+
+        '@typescript-eslint/naming-convention': [
```
Comment (ismail-syed) [inline packages/eslint-plugin/lib/config/typescript.js:73]: Can we note this change in the changelog, please.

**Review by ismail-syed — APPROVED** [review]

**Comment by ismail-syed** [comment]
> Ensure all generic type names start with a T prefix

This is 👍 with me.

> Ensure all interfaces don't start with an I prefix

I personally don't like interfaces that start with `I`, it just looks funny to me so I'm 👍 with this too.

**Comment by atesgoral** [comment]
@kaelig Seems like it!

File: packages/eslint-plugin/lib/config/typescript.js (hunk 1 above)
Comment (atesgoral) [reply packages/eslint-plugin/lib/config/typescript.js:73]: @ismail-syed I'll do it.

**Review by BPScott — APPROVED** [review]
I'm slightly throw as this seems too have two similar but subtly different ideas that I would have expected to be a bit more consistent:

- There is value in using a prefixed-based naming convention to tell the difference between a Generic Type and any other Type is good so we should enforce a prefix
- There is no value in using a prefix-based naming convention to tell the difference between a Type and a concrete value so we should ban using a prefix.


That's probably just me being contrarian though. I don't have a strong opinion on this "banning a thing we already basically never do" is fine by me :shipit:

**Comment by atesgoral** [comment]
@BPScott I don't have a strong opinion on the matter, but if I were to try to justify the combination of these two new rules:

Perhaps duck typing can be seen as interface analogue and therefore we could say "don't prefix JS-native types, whether they're interfaces (duck-typed things) or concrete types implementing those interfaces". On the other hand, a generic type is a preprocessor thing that only belong to the TS realm and so we want to prefix generic types with "T" to make them stand out.

`¯\_(ツ)_/¯`

**Comment by dahukish** [comment]
Not to go against what MS does, and this might be over simplifying things but I think prefixes in general lower the cognitive load when reading someone else's code. To that Effect I would say that we add both the `I` and the `T` or some other naming convention that avoids the assumption that my editor will pickup that it is an interface/type or that I will need to open multiple files. 🤷

**Comment by dahukish** [comment]
@gmcgibbon we good to ship this?

**Pull request merged**